keps/sig-api-machinery/3610-env-injection/README.md

```diff
@@ -0,0 +1,855 @@
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This KEP defines a new annotation `glabel.env` and introduces a mechanism that can override current environment variables on pods.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+This seems to be a common requirement.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- inject env to all containers in namespace
+- skip the env if container already has set
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+Add a new Admission Controller: GlobalEnv. Users can enable or disable it on the apiserver side.
+If enabled, the admission controller will add env according to namespace annotation to every pod's containers in the namespace wide.
+A namespace annotation named `global.env`, works for pods in the namespace.
+
+### Other Proposals and Concerns
+
+**annotation vs configmap**
+Annotation is simpler, So I choose to use an annotation here.
+
+- For annotation, we should divide the string by `,` and get the key and value by `=`, for instance, "key1=value1,key2=value2".
+
+- A configmap named `global.env`. Then we can use the configmap key as env key, and value as env value.
+
+**cluster wide VS namespace wide**
```

Comment:
We did the cluster wide by adding the Admission config file that provides a list of cluster wide env that gets injected into all pods across the namespace with a annotation that would let one opt out of getting env injected into the pod if required.

Reply:
In this design, there would be a new admission config yaml for this feature and apiserver should load it. Is there any admission controller that uses this pattern? I need to check.

Reply:
`PodPreset` used to be able to do this. If I remember correctly.  Where it had its own external config file that you could use ENV values specified right ?